### PR TITLE
Compatibility fix for passing `odc-geo` GeoBoxes to `dc.load(like=...)`

### LINKED ISSUE #1549: Cannot pass `odc.geo` GeoBoxes to `dc.load`
### Expected behaviour
A user may try to use an `odc.geo` GeoBox to load data from `datacube-core`. e.g.:

```
from odc.geo.geobox import GeoBox

gbox = GeoBox.from_bbox(
    [149.036528545, -35.384897602, 149.233835134, -35.198132594], resolution=0.002
)

ds = dc.load(
    product="ga_ls8c_ard_3",
    measurements=["nbart_red", "nbart_green", "nbart_blue"],
    time=("2019-01-10", "2019-01-15"),
    like=gbox,
    resampling="nearest",
    group_by="solar_day",
)
```

### Actual behaviour

However, the code above fails with an `AttributeError`:
image

To make it work correctly, the user must add `.compat` to the end of the `odc.geo` GeoBox to get out an older-style `datacube` GeoBox, e.g. `gbox.compat`:

```
from odc.geo.geobox import GeoBox

gbox = GeoBox.from_bbox(
    [149.036528545, -35.384897602, 149.233835134, -35.198132594], resolution=0.002
)

ds = dc.load(
    product="ga_ls8c_ard_3",
    measurements=["nbart_red", "nbart_green", "nbart_blue"],
    time=("2019-01-10", "2019-01-15"),
    like=gbox.compat,
    resampling="nearest",
    group_by="solar_day",
)
```

Even though our timeline for `odc-geo` integration isn't expected until version 1.9, it would be very helpful to support `odc-geo` GeoBoxes in `dc.load` even in version 1.8 to allow users to start writing code that integrates with `odc-geo`.

This could be as simple as detecting `odc-geo` GeoBoxes and converting them via the `.compat` method internally within `datacube-core`.

### Environment information

* Which ``datacube --version`` are you using? `1.8.17`
* What datacube deployment/enviornment are you running against? `DEA prod Sandbox

> **Note:** Stale issues will be automatically closed after a period of six months with no activity. 
> To ensure critical issues are not closed, tag them with the Github `pinned` tag.
> If you are a community member and not a maintainer please escalate this issue to maintainers via
> [GIS StackExchange](https://gis.stackexchange.com/questions/tagged/open-data-cube) or [Slack](http://slack.opendatacube.org).

### PR DESCRIPTION
**Pull request by robbibt**
### Reason for this pull request
Support users passing `odc.geo` GeoBoxes to `datacube.load(like=...)`. For more details, see #1549.

Temporary fix until `odc-geo` is officially integrated in version 1.9.

### Proposed changes
If the passed object is not a Datacube-style Geobox but has a `odc-geo` ".compat" attribute, try using that ".compat" method to convert the `odc-geo` style GeoBox to a `datacube`-style GeoBox.

(I deliberately didn't use a check for whether it was the object was a `odc.geo.geobox.Geobox` object because I didn't know if we wanted to add `odc-geo` to the package dependencies yet - happy to change to that though if you'd prefer!) 

 - [x] Closes #1549
 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

----
📚 Documentation preview 📚: https://datacube-core--1551.org.readthedocs.build/en/1551/

### CODE REVIEW

**Review by SpacemanPaul — CHANGES_REQUESTED** [review]
Thanks Robbi - would be good to have a test though.

It should be reasonably easy to write a test without installing odc-geo - because of the way you wrote it you could just pass in a mock with an ODC-GEO  `GeoBox` as the `compat` attribute.

File: datacube/api/core.py [hunk 1]
```diff
@@ -875,9 +875,16 @@ def output_geobox(like=None, output_crs=None, resolution=None, align=None,
...
         assert align is None, "'like' and 'align' are not supported together"
+
+        # If user passes a GeoBox, return as-is
         if isinstance(like, GeoBox):
             return like
 
+        # For `odc-geo` GeoBox compatibility: use "compat" attribute if
+        # it exists to convert to a Datacube-style GeoBox
+        if hasattr(like, "compat"):
+            return like.compat
```
Comment (Kirill888) [inline datacube/api/core.py]: I think just having `.compat` is too generic of a check, below grabs an attribute named `.compat` and if that is of `GeoBox` type, use that

```suggestion
        if isinstance(compat := getattr(like, "compat", None), GeoBox):
            return compat
```

Comment (robbibt) [reply datacube/api/core.py]: Looks great to me

**Comment by robbibt** [comment]
@SpacemanPaul Let me know what you think about this approach - I couldn't get MagicMock to work without infinite recursion errors, so went with making a fake class instead.

**Review by SpacemanPaul — APPROVED** [review]
Looks good.  Thanks Robbi (and Kirill)

**Pull request merged**